content/blog/2024/2024-02-13-apache-http-client-config.md

```diff
@@ -0,0 +1,267 @@
+---
+authors: [sagaofsilence]
+title: "Apache HTTP Client Configuration"
+categories: ["Java"]
+date: 2024-02-11 00:00:00 +1100
+excerpt: "Apache HTTP Client Configuration."
+image: images/stock/0125-tools-1200x628-branded.jpg
+url: apache-http-client-config
+---
+
+Configuring Apache HTTP Client is essential for tailoring its behavior to meet specific requirements and optimize performance. From setting connection timeouts to defining proxy settings, configuration options allow developers to fine-tune the client's behavior according to the needs of their application. In this section, we will explore various configuration options available in Apache HTTP Client, covering aspects such as connection management, request customization, authentication, and error handling. Understanding how to configure the client effectively empowers developers to build robust and efficient HTTP communication within their applications.
+
+## The "Create a HTTP Client with Apache HTTP Client" Series
+
+This article is the second part of a series:
+
+1. [Introduction to Apache HTTP Client](/create-a-http-client-with-apache-http-client/)
+2. [Apache HTTP Client Configuration](/apache-http-client-config/)
+3. [Classic APIs Offered by Apache HTTP Client](/apache-http-client-classic-apis/)
+4. [Async APIs Offered by Apache HTTP Client](/apache-http-client-async-apis/)
+5. [Reactive APIs Offered by Apache HTTP Client](/apache-http-client-reactive-apis/)
+
+{{% github "https://github.com/thombergs/code-examples/tree/master/create-a-http-client-wth-apache-http-client" %}}
+
+\
+Let us now learn commonly used options to configure Apache HTTP client for web communication.
+
+## HttpClient Client Connection Management
+Connection management in Apache HTTP Client refers to the management of underlying connections to remote servers. Efficient connection management is crucial for optimizing performance and resource utilization. Apache HTTP Client provides various options for configuring connection management.
+
+`PoolingHttpClientConnectionManager` manages a pool of client connections and is able to service connection requests from multiple execution threads. Connections are pooled on a per route basis.
+
+`PoolingHttpClientConnectionManager` maintains a maximum limit of connections on a per route basis and in total. By default it creates up to 2 concurrent connections per given route and up to 20 connections in total. For real-world applications we cam increase these limits if needed.
+
+This example shows how the connection pool parameters can be adjusted:
+
+```java
+  public CloseableHttpClient getPooledCloseableHttpClient(
+      String host,
+      int port,
+      int maxTotalConnections,
+      int defaultMaxPerRoute,
+      long requestTimeoutMillis,
+      long responseTimeoutMillis,
+      long connectionKeepAliveMillis) {
+    PoolingHttpClientConnectionManager connectionManager =
```

Comment:
Let's move the parentheses to a new newline and also add a blank line after each line of code that contain a line break. That makes to code more readable.

Reply:
Formatted code examples to make those more readable.